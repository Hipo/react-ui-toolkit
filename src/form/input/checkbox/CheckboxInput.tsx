import CheckIcon from "../../../ui/icons/check.svg";

import "./_checkbox-input.scss";

import React from "react";
import classNames from "classnames";

export interface CheckboxInputItem<Id = string, Context = any> {
  id: Id;
  content: React.ReactNode;
  inputProps: {
    htmlFor: string;
    value: string;
    name: string;
  };
  context?: Context;
}

export interface CheckboxInputProps {
  item: CheckboxInputItem;
  onSelect: (
    item: CheckboxInputItem,
    event?: React.SyntheticEvent<HTMLInputElement>
  ) => void;
  isSelected: boolean;
  isDisabled?: boolean;
  customClassName?: string;
  testid?: string;
}

function CheckboxInput({
  item,
  onSelect,
  customClassName,
  isSelected,
  isDisabled,
  testid
}: CheckboxInputProps) {
  const {
    inputProps: {name, value, htmlFor},
    content
  } = item;
  const containerClassName = classNames("checkbox-input-label", customClassName, {
    "checkbox-input-label--is-selected": isSelected,
    "checkbox-input-label--is-disabled": isDisabled
  });

  return (
    <label data-testid={testid} htmlFor={htmlFor} className={containerClassName}>
      <input
        type={"checkbox"}
        id={htmlFor}
        name={name}
        value={value}
        className={"checkbox-input"}
        onChange={handleCheckboxChange}
        checked={isSelected}
        disabled={isDisabled}
      />

      <span className={"checkbox-input-label__icon"}>
        <CheckIcon />
      </span>

      {content}
    </label>
  );

  function handleCheckboxChange(event: React.SyntheticEvent<HTMLInputElement>) {
    onSelect(item, event);
  }
}

export default CheckboxInput;
