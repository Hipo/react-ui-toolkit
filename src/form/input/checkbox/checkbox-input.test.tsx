import React from "react";
import {render, cleanup, fireEvent} from "@testing-library/react";
import "@testing-library/jest-dom";
import {create} from "react-test-renderer";

import CheckboxInput, {CheckboxInputProps} from "./CheckboxInput";
import {testA11y} from "../../../core/utils/test/testUtils";

describe("<CheckboxInput />", () => {
  afterEach(cleanup);

  const defaultCheckboxInputProps: CheckboxInputProps = {
    testid: "checkbox-input",
    isSelected: false,
    item: {
      id: "checkbox-input",
      content: <p data-testid={"checkbox-input.content"}>{"Test"}</p>,
      inputProps: {name: "checkbox-input", htmlFor: "checkbox-input", value: "Test"}
    },
    onSelect: jest.fn()
  };

  it("should render correctly", () => {
    render(<CheckboxInput {...defaultCheckboxInputProps} />);
  });

  it("should matches snapshot", () => {
    const tree = create(<CheckboxInput {...defaultCheckboxInputProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should pass a11y test", async () => {
    const {container} = render(<CheckboxInput {...defaultCheckboxInputProps} />);

    await testA11y(container);
  });

  it("disabled property should works correctly", () => {
    const {getByTestId} = render(
      <CheckboxInput isDisabled={true} {...defaultCheckboxInputProps} />
    );

    expect(getByTestId("checkbox-input")).toHaveClass(
      "checkbox-input-label--is-disabled"
    );

    const checkboxInput = document.getElementsByClassName("checkbox-input")[0];

    expect(checkboxInput).toHaveAttribute("disabled");
    expect(checkboxInput).toBeDisabled();
  });

  it("isSelected property should works correctly", () => {
    const {getByTestId, rerender} = render(
      <CheckboxInput {...defaultCheckboxInputProps} />
    );

    const checkboxInput = document.getElementsByClassName("checkbox-input")[0];

    expect(checkboxInput).not.toBeChecked();

    rerender(<CheckboxInput {...defaultCheckboxInputProps} isSelected={true} />);

    expect(checkboxInput).toBeChecked();
    expect(getByTestId("checkbox-input")).toHaveClass(
      "checkbox-input-label--is-selected"
    );
  });

  it("should call onSelect prop correctly", () => {
    render(<CheckboxInput {...defaultCheckboxInputProps} />);

    const checkboxInput = document.getElementsByClassName("checkbox-input")[0];

    fireEvent.click(checkboxInput, {target: {checked: true}});

    expect(defaultCheckboxInputProps.onSelect).toHaveBeenCalledTimes(1);
  });
});
