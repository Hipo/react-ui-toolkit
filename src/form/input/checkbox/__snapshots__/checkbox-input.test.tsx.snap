// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxInput /> should matches snapshot 1`] = `
<label
  className="checkbox-input-label"
  data-testid="checkbox-input"
  htmlFor="checkbox-input"
>
  <input
    checked={false}
    className="checkbox-input"
    id="checkbox-input"
    name="checkbox-input"
    onChange={[Function]}
    type="checkbox"
    value="Test"
  />
  <span
    className="checkbox-input-label__icon"
  >
    <svg />
  </span>
  <p
    data-testid="checkbox-input.content"
  >
    Test
  </p>
</label>
`;
